icu4c/source/common/unifiedcache.h

```diff
@@ -146,31 +155,9 @@ class LocaleCacheKey : public CacheKey<T> {
    virtual int32_t hashCode() const {
        return (int32_t)(37u * (uint32_t)CacheKey<T>::hashCode() + (uint32_t)fLoc.hashCode());
    }
-   virtual bool operator == (const CacheKeyBase &other) const {
-       // reflexive
-       if (this == &other) {
-           return true;
-       }
-       if (!CacheKey<T>::operator == (other)) {
-           return false;
-       }
-       // We know this and other are of same class because operator== on
-       // CacheKey returned true.
-       const LocaleCacheKey<T> *fOther =
-               static_cast<const LocaleCacheKey<T> *>(&other);
-       return fLoc == fOther->fLoc;
+   inline bool operator == (const LocaleCacheKey<T> &other) const {
+       return fLoc == other.fLoc;
```

Comment:
I think it works because, for our internal usage, there aren't any possibly-conflicting subclasses. 
`class DateFmtBestPatternKey : public LocaleCacheKey<DateFmtBestPattern>` comes closest, but there aren't any other uses of `LocaleCacheKey<DateFmtBestPattern>`.

`LocaleCacheKey<>`, templatized on different types, are completely separate types, so they won't conflict.

But this stuff is sooo confusing, I could easily be overlooking something.